<a href="https://colab.research.google.com/github/Khanbir/Data_analitycs/blob/main/DZ_14_2_%D0%9F%D0%B5%D1%80%D0%B5%D0%B2%D1%96%D1%80%D0%BA%D0%B0_%D1%81%D1%82%D0%B0%D1%82%D0%B8c%D1%82%D0%B8%D1%87%D0%BD%D0%B8%D1%85_%D0%B3%D1%96%D0%BF%D0%BE%D1%82%D0%B5%D0%B7_Z_test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Перевірка статиcтичних гіпотез. Z-test.



0. Зчитайте дані з `data.csv` у змінну data, яка має тип pandas.DataFrame. Ми будемо далі працювати з цією змінною.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [7]:
import pandas as pd
data = pd.read_csv('/content/drive/MyDrive/data.csv')
print(data)

    Unnamed: 0       data
0            0  17.499453
1            1  19.662399
2            2   7.182823
3            3  29.841625
4            4   9.239386
..         ...        ...
95          95  26.244944
96          96  22.282832
97          97  18.697188
98          98  27.041189
99          99   5.938381

[100 rows x 2 columns]


Запустіть код нижче. Для коректної робити всіх подальших методів дані мають бути у вигляді одновимірного `numpy.ndarray` вектора та мати розмірність (100,). Така розмірність означає, що дані - одновимірні (колонка), якщо б розмірність була (100, 1), то дані сприймаються методами як двовимірні (таблиця), хоча для нас виглядати можуть так само.

In [3]:
data = data.data.values
data.shape

(100,)

**Завдання 1**. Зчитані дані - це сума покупок в доларах наших покупців на сайті протягом липня 2025 року.

До цього місяця, ми мали продажі в середньому на чек $20.

Необхідно зʼясувати, чи вийшло у нас статистично значущо **підвищити** середній чек за липень 2025?

Аби дати відповідь - ми проведемо z-test на рівні значущості $\alpha=0.05$ двома способами. В цьому завданні ми будемо виконувати обчислення "вручну" з використанням бібліотек numpy та scipy.stats подібно до прикладу в лекції. Для цього виконайте наступні 7 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть параметри задачі у змінні Python:
    - `sample_mean` - $\bar{x}$, середнє значення в вибірці
    - `population_mean` - $\mu_0$, середнє значення в популяції (тобто те, з яким ми порівнюємо середнє вибіркове значення)
    - `population_std` - $\sigma$, вибіркове стандартне відхилення, яке ми вважаємо, що є рівним ст. відх. популяції, адже маємо достатньо велику вибірку.
    - `sample_size` - $n$, розмір вибірки
    - `alpha` - $\alpha$ рівень значущості
    
    **Увага!** Для обчислення стандартного відхилення маємо скористатись функцією `np.std(your_dataframe, ddof=1)`. Чому так - розбираємо в лекції про t-test.


2. Визначте для цієї задачі:
    - якою є гіпотеза $H_0$
    - якою є альтернативна гіпотеза $H_a$
    - з яким типом тесту ми маємо справу - лівосторонній, правосторонній чи двосторонній.

3. Обчисліть стандартну помилку SE.
4. Розрахуйте z-статистику.
5. Знайдіть критичне z-значення з допомогою бібліотеки stats.
6. Обчисліть p-value з допомогою бібліотеки stats.
7. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$. Для прийняття рішення зробіть і порівняння z-статистики з критичним значенням, і проаналізуйте p-value.

In [10]:
import pandas as pd
import numpy as np
from scipy import stats

# Середнє значення в нашій вибірці (липневі продажі)
sample_mean = data['data'].mean()

# Середнє значення в популяції (середній чек до липня 2025)
population_mean = 20

# Стандартне відхилення вибірки (вважаємо рівним стандартному відхиленню популяції)
# ddof=1 для вибіркового стандартного відхилення
population_std = data['data'].std(ddof=1)

# Розмір вибірки
sample_size = len(data)

# Рівень значущості
alpha = 0.05

print(f"Середнє значення вибірки (x̄): {sample_mean:.2f}")
print(f"Середнє значення популяції (μ₀): {population_mean}")
print(f"Стандартне відхилення (σ): {population_std:.2f}")
print(f"Розмір вибірки (n): {sample_size}")
print(f"Рівень значущості (α): {alpha}");

se = population_std / np.sqrt(sample_size)
print(f"Стандартна помилка (SE): {se:.2f}");

z_statistic = (sample_mean - population_mean) / se
print(f"Z-статистика: {z_statistic:.2f}");

critical_z = stats.norm.ppf(1 - alpha)
print(f"Критичне z-значення: {critical_z:.2f}");

p_value = 1 - stats.norm.cdf(z_statistic)
print(f"P-value: {p_value:.4f}")

Середнє значення вибірки (x̄): 19.38
Середнє значення популяції (μ₀): 20
Стандартне відхилення (σ): 6.25
Розмір вибірки (n): 100
Рівень значущості (α): 0.05
Стандартна помилка (SE): 0.63
Z-статистика: -1.00
Критичне z-значення: 1.64
P-value: 0.8402


Аналіз та висновки
Порівняння Z-статистики та критичного Z-значення
Ваша z-статистика дорівнює -1.00, а критичне z-значення — 1.64. У правосторонньому тесті ми відхиляємо нульову гіпотезу, якщо z-статистика більша за критичне значення.

Z статистика ​(−1.00) !> Zкритичне(1.64)

Оскільки −1.00 не перевищує 1.64, z-статистика не потрапляє в область відхилення нульової гіпотези. Це говорить нам про те, що немає достатніх доказів для відхилення (H0).

Аналіз P-value
Ваше p-value становить 0.8402, а рівень значущості (α) — 0.05. Для прийняття рішення ми порівнюємо ці два значення.

p-value(0.8402)>α(0.05)

Оскільки ваше p-value значно більше за рівень значущості, це вказує на те, що ймовірність отримати такий результат (або ще більш екстремальний) випадково є дуже високою, якщо нульова гіпотеза правдива.

Остаточне рішення
На підставі обох порівнянь (z-статистики з критичним значенням та p-value з α), ми не маємо підстав відхиляти нульову гіпотезу (H0).

Висновок: Незважаючи на те, що ми хотіли підвищити середній чек, за наданими даними ми не можемо стверджувати, що відбулося статистично значуще підвищення. Середній чек у липні (19.38) фактично виявився меншим за середній показник популяції (20), що є ще одним доказом на користь того, що ціль не була досягнута.

**Завдання 2.** Виконайте обчислення z-test з використанням бібліотеки statsmodels.

Отримайте z-статистику та р-значення.

Виведіть p-значення та зробіть висновок, чи ми маємо достатньо доказів, аби стверджувати, що середній чек зріс.

Чи зійшлись значення z-статистику та р-значення в цьому завданні з попередніми обчисленнями?

Завдання оцінюється в 3 бали.

In [12]:
import pandas as pd
from statsmodels.stats.weightstats import ztest

# Припустимо, що дані знову зчитані
sample = data['data']

# Виконуємо z-тест
z_statistic, p_value = ztest(x1=sample, value=20, alternative='larger')

print(f"Z-статистика: {z_statistic:.2f}")
print(f"P-value: {p_value:.4f}")

Z-статистика: -1.00
P-value: 0.8402


Висновок
Отримані значення:

Z-статистика: -1.00

P-value: 0.8402

Оскільки p-value (0.8402) більше, ніж рівень значущості
alpha (0.05), ми не відхиляємо нульову гіпотезу. Це означає, що немає достатньо статистичних доказів, аби стверджувати, що середній чек статистично значуще зріс.

Порівняння з попередніми обчисленнями
Значення z-статистики (-1.00) та p-значення (0.8402), отримані за допомогою бібліотеки statsmodels, повністю збігаються зі значеннями, які були розраховані "вручну" в попередньому завданні. Це підтверджує правильність обох методів обчислення. Використання бібліотек значно спрощує процес, автоматизуючи всі необхідні розрахунки(2 строчки замість 10)!!!